### ISSUE TITLE
Sub-optimal performance of small model and a question on quantization

### ISSUE THREAD

**Issue by aclex** [kendryte/nncase#294]
I continue my experiments with trying to bring some small model all the way from PyTorch implementation to Kmodel using the fresh ONNX importing backend (plan to PR it as an example when done), and yesterday finally succeeded in getting it to work (that's indeed looked like wonder! I'm still happy with that). But I noticed the performance seems to be quite sub-optimal for such a tiny model. Here's the code in PyTorch to have an idea of the structure:

```python
class Net(nn.Sequential):
	def __init__(self):
		super(Net, self).__init__(
			nn.Conv2d(1, 32, 3, 1),
			nn.ReLU(),
			nn.Conv2d(32, 64, 3, 1),
			nn.ReLU(),
			nn.MaxPool2d(4),
			nn.Conv2d(64, 10, 3, 1), # to workaround the problem with nn.Linear() described in #288
			nn.AvgPool2d(4),
			nn.Flatten(),
			nn.Softmax(dim=1)
		)

	def forward(self, x):
		return torch.log(super().forward(x))
```

The shape of the input is `(1, 1, 28, 28)` i.e. one channel 28×28 raw picture. And the processing time, according to NNDEBUG output, is about 128 ms on K210:
```bash
run kmodel takes 129.057007 ms.
```

Generally it's OK, but, given it used to show me more, than 70 FPS with facedetect model, I'm thinking on the ways of improvement.

One particular question I have on the quantization: is it better to have the model pre-quantized using ONNX or PyTorch or something before fetching it to `ncc`, given it has its very own set of transforms, or `ncc` is just enough to compress the normal `float32` model? I mean, I noticed that even on late stages (`target_dependent_after_buffer_fusion`) the convolutions seem to be floating-point and just folded in quantize/dequantize ops, which lets me thinking it slows down the process. Is it at all possible to get integer convolutions and other operations? If possible, is it better? There seems to be `ConvInteger` and `MatMulInteger` ops in ONNX, maybe, we need them to have these really quantized?

Any help would be very appreciated.

**Comment by aclex** [comment]
Uploaded all the results to https://github.com/aclex/k210_mnist_assets.

**Comment by sunnycase** [comment]
Generally PTQ(post-training quantization) models will have better performance than QAT(quantize-aware training) models. Because QAT models already fuse Convs with Acts, nncase cannot assume whether there are Acts or not after Convs. It will  cause nncase to disable some optimization transforms.

**Comment by aclex** [comment]
Many thanks for clarification! This makes sense.

So, in terms of destination model graph, are floating-point convolutions (and matmuls, maybe) wrapped with quantize-dequantize and table lookup operators harmful for on-device performance and should be avoided, compared to integer convolutions (and possibly matmul), or they are just OK? I mean, are they parsed on hardware to something fast or it just has to perform (usually quite slow) floating point calculations?

**Comment by sunnycase** [comment]
float Convs and MatMuls should be converted to `kpu_conv2d` before `target_dependent_after_buffer_fusion` stage or they will be executed slowly on cpu.

**Comment by aclex** [comment]
Thanks again for all the details on that! That basically matches my assumptions. I've also revisited the model after buffer fusion and indeed found calls to the integer KPUConv2d operators there, which means it works as expected. Unfortunately, after the recent updates in nncase and SDK this particular model stopped working for me on K210, but, I think, I'll sort this out.

**Comment by aclex** [comment]
I decided to reopen this issue to try to sort these performance issues out. Could you please help me with this? The problem is just it: this little model (uploaded at https://github.com/aclex/k210_mnist_assets) runs at ~130 ms on K210 with 28×28×1 picture, which looks like quite a low performance. At first glance, the optimized and quantized model (after buffer fusion) looks fine, as it performs integer operations mostly. The only concerns I see is float32 `Pad` and `Quantize`-`Dequantize` pairs all over the place. Could you please take a look at this? You surely have more experience with it and maybe see the problem quickly.

My ultimate goal is to get a handful path of converting bigger models (e.g. MobileNetv3) from PyTorch to Kmodel with proper performance, I saw there's already a test with MobileNetv2 converted from tflite and example with YOLOv5 from Caffe, so I decided to start with something very simple and stuck a little bit with this performance issue.

**Comment by aclex** [comment]
So I've analyzed models a little bit more and found that up til the `Optimize target dependent` stage the model looks quite expected in terms of optimization and efficiency:
![After optimization](https://raw.githubusercontent.com/aclex/k210_mnist_assets/master/after_indep.svg "After optimization")
While after `Add quantize annotation` stage additional `Pad` nodes appear:
![After quantize annotation](https://raw.githubusercontent.com/aclex/k210_mnist_assets/master/after_quant_annot.svg "After quantize annotation")

Now basically the only question appears is why the `Pad`, `ReduceWindow2d` and `Concat` (found in YOLOX model in examples) IR nodes are not quantized and converted to integer calculation, but instead require KPU upload/download, dequantize and quantize for them to work, which is obviously what hurt the performance most of all.

**Comment by sunnycase** [comment]
@aclex Please take a look at https://github.com/kendryte/nncase/blob/master/docs/FAQ_EN.md
>2. What kinds of ops can be fully accelerated by KPU?
    Same symmetric paddings (TensorFlow use asymmetric paddings when stride=2 and size is even).

You should use symmetric paddings to avoid additional `Pad`.

**Comment by aclex** [comment]
@sunnycase Yes, sure, this is my every day document now :) So the padding is **required** to be of the value to preserve the same size? I seem to misunderstood this point. I mean, there's no padding set in the convolutions in the source model now whatsoever (`pads` attribute is `0, 0, 0, 0` for all the `Conv` ops and one remaining `Pad` op is also fully zero and eliminated during target-independent optimization step. I just thought, it would be even better with no padding, than with any non-zero padding. Thank you so much for pointing me out! Will try the model with padding this evening.

**Comment by aclex** [comment]
@sunnycase thanks again for your kind reference to the padding settings! I've just checked the model with the padding of 1 and now it's compiled almost perfectly right! Only the ending part of `Flatten` and `Softmax` and `Log` are calculated in floats:
![Fastest model for now](https://raw.githubusercontent.com/aclex/k210_mnist_assets/master/after_buf_fusion.svg "After buffer fusion")
The inference time is now also much less — only ≈10.2 ms! I think, it's near the maximum for this model and generally I'm fully sure to close this, but still wanted to ask finally: could those unary remaining ops like softmax and log be optimized and calculated in integer or they always "emulated" on the CPU? Just out of interest, I'm still satisfied with the speed for now.

**Comment by sunnycase** [comment]
We will not add more ops to `stackvm` module.
Instead we are planning to add a new `cpu aot` module which will embed target specific and pre-compiled machine code to executed optimized ops such as softmax etc.

**Comment by aclex** [comment]
Many thanks for clarification!
